### PR TITLE
Add step-88: matrix-free, non-nested, geometric multigrid

### PR DESCRIPTION
**Pull request by fdrmrc**
This PR, written by @peterrum and me, adds a tutorial which shows the usage of the non-nested multigrid infrastructure. It solves a Poisson problem on a sequence of non-nested meshes of a piston, stemming out of CAD files.

### CODE REVIEW

**Comment by blaisb** [comment]
@fdrmrc  Great example. I'll gladly participate in reviewing it. Quick question, how did you generate the coarse mesh of the piston ? It's a nice looking unstructured hex mesh.

**Comment by fdrmrc** [comment]
Thanks!
> Quick question, how did you generate the coarse mesh of the piston ? It's a nice looking unstructured hex mesh.

I started with a CAD model of a piston and, using Cubit, repaired and meshed it manually.

File: examples/step-88/doc/intro.dox [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+- We create the meshes by globally refining a coarse mesh. This results in
+  a nested mesh and therefore does not require MGTwoLevelTransferNonNested.
+  However, this gives a good comparison to MGTwoLevelTransfer in terms of
+  convergence rate and performance.
+
+
+<h3>%Simplex meshes</h3>
+
+This tutorial works for hyper-cube and simplex meshes. In the case
+of simplex meshes, different mapping, finite elment, and quadrature classes
```
Comment (blaisb) [inline examples/step-88/doc/intro.dox:122]:
```suggestion
of simplex meshes, different mapping, finite element, and quadrature classes
```

**Review by blaisb — CHANGES_REQUESTED** [review]
I have begun reviewing. Will finish this weekend.

**Comment by drwells** [comment]
We've run out of time to add this to 9.7 since we want to tag a release soon and we haven't done any significant review yet.

**Comment by bangerth** [comment]
What's the status here? Do we want to shoot for the upcoming release? (@blaisb You seem to have started reviewing the program. Do you want to continue with that?)

**Comment by bangerth** [comment]
@fdrmrc and @blaisb Any news?

**Comment by bangerth** [comment]
Last call for the upcoming release!

**Comment by blaisb** [comment]
Sorry, I missed the boat on this one. I can provide a review by next weekend I believe, but it would be relevant if someone else than me also reviewed.

File: examples/step-88/doc/intro.dox [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    </td>
+  </tr>
+</table>
+
+The prolongation operation (for interpolation from coarse grid to fine
+grid) is given by the evaluation of finite element solutions at support points
+of the fine mesh within a coarse mesh. How this can be accomplished with
+Utilities::MPI::RemotePointEvaluation and VectorTools::point_values() is
+shown in step-87. The restriction operation is naturally given by the
+transposed of the prolongation operation.
```
Comment (bangerth) [inline examples/step-88/doc/intro.dox:75]:
```suggestion
transpose of the prolongation operation.
```

File: examples/step-88/doc/intro.dox [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+<h3>The test case</h3>
+
+For demonstration purposes only, we will solve a classical Poisson problem with
+Dirichlet boundary conditions:
+@f{align*}{
+  -\Delta u &= 1 \qquad && \text{in }\, \Omega,
+  \\
+  u &= u_D       \qquad && \text{on }\, \Gamma = \partial \Omega,
+@f}
+on a given domain $\Omega$.
```
Comment (bangerth) [inline examples/step-88/doc/intro.dox:101]: Perhaps add something like the following:

```suggestion
For demonstration purposes only, we will solve a classical Poisson problem with
Dirichlet boundary conditions:
@f{align*}{
  -\Delta u &= 1 \qquad && \text{in }\, \Omega,
  \\
  u &= u_D       \qquad && \text{on }\, \Gamma = \partial \Omega,
@f}
on a given domain $\Omega$. One could think of this as solving the problem of heat generation and
conduction within the object shown above, where $u$ represents the temperature, $u_D$ the
temperature at the boundary, and there is unit heat production per volume inside the object (for
example through radioactive decay or through electromagnetic heating). In most mechanical
applications, heat isn't produced *within* a solid body, but is transferred to it from the outside
(in the example of the piston geometry here, heat would be produced by the combusting fuel),
but this tutorial is about setting up non-matching multigrid, not a physically realistic situation, and
so the simple equation above will do to showcase how this can be achieved.
```

File: examples/step-88/doc/intro.dox [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+
+<h3>%Mesh sequence</h3>
+
+In this tutorial, we consider two types of mesh sequences:
+- We generate $\Omega$ out of a CAD model of a <i> piston </i> with the
+external mesh generator software Coreform CUBIT. The required hierarchy is
+generated during a pre-processing step inside the mesh generator by varying
+iteratively the mesh-size for each grid and finally exporting it in a format that
+can be read by the GridIn class. Such mesh files are given in the grid/ folder of
```
Comment (bangerth) [inline examples/step-88/doc/intro.dox:111]:
```suggestion
can be read by the GridIn class. Such mesh files are given in the `grids/` folder of
```

File: examples/step-88/doc/intro.dox (hunk 1 above)
Comment (bangerth) [inline examples/step-88/doc/intro.dox:122]: I think you want to say it as follows:

```suggestion
This tutorial works uses meshes that consist of hyper-cube cells. However, everything in this
program also works if you used simplex meshes though in that case
different mapping, finite element, and quadrature classes
```

File: examples/step-88/doc/results.dox [hunk 5]
```diff
@@ -0,0 +1,76 @@
+<h1>Results</h1>
+
+The directory in which this program is run contains a sample parameter file
+that you can use to reproduce the results presented in this
+section. Its default values are listed as follows:
```
Comment (bangerth) [inline examples/step-88/doc/results.dox:5]:
```suggestion
section. This file will be read by the ParameterHandler class; in contrast to other
tutorial programs, it doesn't use the traditional `.prm` file format, but is a JSON file.
Its default values are listed as follows:
```

File: examples/step-88/doc/results.dox [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  "Dimension": 3,
+  "NGlobalRefinements": 3,
+  "Degree": 2,
+  "SolverMaxIterations": 100,
+  "SolverAbsTolerance": 1e-20,
+  "SolverRelTolerance": 1e-4,
+  "MGSmoothingRange": 20,
+  "MGSmootherDegree": 5,
+  "MGSmootherEigNIterations": 20,
+  "MGNonNested": true
```
Comment (bangerth) [inline examples/step-88/doc/results.dox:20]: Might be nice to align the values for easier readability.

File: examples/step-88/doc/results.dox [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+Solved in 3 iterations.
+@endcode
+
+
+The output to Paraview is as follows:
+
+<table align="center" class="doxtable">
+  <tr>
+    <td>
+        @image html piston_contour.png
```
Comment (bangerth) [inline examples/step-88/doc/results.dox:54]: This line will have to be changed to link to where all other tutorials store their files. Take a look at any other tutorial how they link to images.

Comment (bangerth) [reply examples/step-88/doc/results.dox:54]: I added the file to the repository from which we create the website using the name `step-88-piston-solution.png`.

File: examples/step-88/doc/results.dox [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    </td>
+  </tr>
+</table>
```
Comment (bangerth) [inline examples/step-88/doc/results.dox:57]: I think it would be nice to provide some context or interpretation here. For example, 3 iterations is really good. It's worth saying that. But then you also mention in the introduction that the program compares the non-nested and a nested hierarchy. (I haven't read the actual program yet, so I don't know whether that's true.) It would be nice to come back to this here: How do the two methods compare in terms of number of iterations, run time, etc? 

I suspect that you have much of that in your paper. I think it would be good to have a brief summary here, and then perhaps reference the paper again for more data.

File: examples/step-88/doc/results.dox [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+<h3>Possibilities for extensions</h3>
+
+This program highlights some of the main capabilities
+of the non-nested-multigrid routines in deal.II. However, there are many
+related topics worth mentioning:
+- Compare the results with MGTwoLevelTransferNonNested with the ones
+  of MGTwoLevelTransfer. For this purpose, set `mg_non_nested = false` and
+  `mesh_type = "hyper_cube"` or `mesh_type = "hyper_cube_with_simplices"`.
```
Comment (bangerth) [inline examples/step-88/doc/results.dox:66]: If the program can also do that without additional changes, then I would present this comparison above. This isn't an "extension" in that case, but already implemented :-)

File: examples/step-88/doc/results.dox [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+- It is possible to run the program with other, user-generated meshes by
+  adjusting the parameter "mesh_file_format". Feel free to experiments with
```
Comment (bangerth) [inline examples/step-88/doc/results.dox:68]:
```suggestion
  adjusting the parameter "mesh_file_format". Feel free to experiment with
```

File: examples/step-88/doc/results.dox [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  meshes with curvature. Here, it is critical that all points of the fine
+  mesh are found on the coarse mesh. For this purpose, one might need
+  to adjust the parameters in MGTwoLevelTransferNonNested::AdditionalData.
+- The tutorial could be extend so that it works for mixed meshes.
```
Comment (bangerth) [inline examples/step-88/doc/results.dox:72]:
```suggestion
- The tutorial could be extended so that it works for mixed meshes, i.e., meshes that have both
  hypercube and simplex cells. Such meshes are common in industrial practice where mesh generators
  resolve complex boundary features with simplices, and then transition to hypercubes in the interior
  of the domain given that they provide better accuracy than simplices.
```

File: examples/step-88/parameters.json [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+  "MeshType": "mesh_file",
+  "MeshFileFormat": "./grids/piston_%d.inp",
+  "Dimension": 3,
+  "NGlobalRefinements": 3,
+  "Degree": 2,
+  "SolverMaxIterations": 100,
+  "SolverAbsTolerance": 1e-20,
+  "SolverRelTolerance": 1e-4,
+  "MGNonNested": true
+}
```
Comment (bangerth) [inline examples/step-88/parameters.json:11]: You'll need an endline character at the end of the file.

File: tests/examples/step-88.output
Comment (bangerth) [inline tests/examples/step-88.output:1]: Is this file still missing?

**Review by bangerth — COMMENTED** [review]
I read through introduction and results. I'll get back to the program at some other time, hopefully not too far in the future.

File: examples/step-88/step-88.cc [hunk 13]
```diff
@@ -0,0 +1,898 @@
...
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE.md at
+ * the top level directory of deal.II.
+ *
+ * ---------------------------------------------------------------------
+ *
+ * Authors: Marco Feder, SISSA, 2024
+ *          Peter Munch, Technical University of Berlin, 2024
```
Comment (bangerth) [inline examples/step-88/step-88.cc:17]: Could you update this to 2026? I'm sorry we let that drag out for so long :-(

File: examples/step-88/step-88.cc [hunk 14]
```diff
@@ -0,0 +1,898 @@
...
+ * Copyright (C) 2024 by the deal.II authors
+ *
+ * This file is part of the deal.II library.
+ *
+ * The deal.II library is free software; you can use it, redistribute
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE.md at
+ * the top level directory of deal.II.
```
Comment (bangerth) [inline examples/step-88/step-88.cc:12]: This also needs the new copyright header. Take a look at any of the other programs.

File: examples/step-88/step-88.cc [hunk 15]
```diff
@@ -0,0 +1,898 @@
...
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_in.h>
+#include <deal.II/grid/grid_tools.h>
+
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/trilinos_precondition.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+#include <deal.II/lac/trilinos_sparsity_pattern.h>
```
Comment (bangerth) [inline examples/step-88/step-88.cc:48]: Since you're using Trilinos, you need to also check for Trilinos in `CMakeLists.txt` (where right now you only check for MPI).

File: examples/step-88/step-88.cc [hunk 16]
```diff
@@ -0,0 +1,898 @@
...
+// include headers related to the FiniteElement space we want to work with, to
+// the TrilinosWrappers for the coarse grid solver, to MatrixFree tools for
+// operator evaluation and to Multigrid infrastructure.
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/mpi.h>
+#include <deal.II/base/parameter_handler.h>
+
+#include <deal.II/distributed/fully_distributed_tria.h>
+#include <deal.II/distributed/repartitioning_policy_tools.h>
+#include <deal.II/distributed/shared_tria.h>
```
Comment (bangerth) [inline examples/step-88/step-88.cc:32]: Do you really use both fully distributed and shared triangulations?

File: examples/step-88/step-88.cc [hunk 17]
```diff
@@ -0,0 +1,898 @@
...
+    // Specify whether the nested or non-nested global-coarsening algorithm
+    // should be used. Note: in the case of "mesh_file", only
+    // a non-nested algorithm can be selected.
+    bool mg_non_nested;
+
+    // Constructor, which sets the default values of the parameters.
+    Parameters();
+
+    // Parse input file.
+    void parse(const std::string &file_name);
```
Comment (bangerth) [inline examples/step-88/step-88.cc:117]: Tutorial programs are often hard to read if commentary and code go back and forth too often. From a readability perspective, it's better to have larger comments commenting on larger code blocks. Would you mind condensing things a bit? A comment can comment on several parameters.

File: examples/step-88/step-88.cc [hunk 18]
```diff
@@ -0,0 +1,898 @@
...
+    prm.add_parameter("Degree", fe_degree);
+
+    prm.add_parameter("SolverMaxIterations", solver_max_iterations);
+    prm.add_parameter("SolverAbsTolerance", solver_abs_tolerance);
+    prm.add_parameter("SolverRelTolerance", solver_rel_tolerance);
+
+    prm.add_parameter("MGNonNested", mg_non_nested);
+
+    std::ifstream file;
+    file.open(file_name);
```
Comment (bangerth) [inline examples/step-88/step-88.cc:166]:
```suggestion
    std::ifstream file(file_name);
    AssertThrow(file.fail() == false, ExcIO());

```

File: examples/step-88/step-88.cc [hunk 19]
```diff
@@ -0,0 +1,898 @@
...
+
+  // Get name of the mesh on the given level. The function replaces "%d"
+  // in the mesh-name format.
+  std::string Parameters::get_mesh_file_name(const unsigned int level) const
+  {
+    char buffer[100];
+
+    std::snprintf(buffer, 100, mesh_file_format.c_str(), level);
+
+    return {buffer};
```
Comment (bangerth) [inline examples/step-88/step-88.cc:179]: I would prefer to not use the old-fashioned `snprintf` function. (In fact, I removed a similar call just the other day in #19760, see also #19761.) I think a better approach would be if instead of a format, you'd store a variable `mesh_file_base_name` and then here you'd do

```suggestion
    return mesh_file_base_name + std::to_string(level) + ".inp";
```

File: examples/step-88/step-88.cc [hunk 20]
```diff
@@ -0,0 +1,898 @@
...
+  }
+
+
+
+  // @sect3{Laplace operator}
+  // A basic matrix-free implementation of the Laplace operator. For
+  // more details, see step-75.
+  template <int dim, typename number>
+  class LaplaceOperator : public Subscriptor
```
Comment (bangerth) [inline examples/step-88/step-88.cc:188]:
```suggestion
  // A basic matrix-free implementation of the Laplace operator. This class
  // implements the basic interface necessary to use it as an operator in a
  // matrix-free context to solve linear systems with the conjugate gradient
  // method. For more details, see step-75.
  template <int dim, typename number>
  class LaplaceOperator : public Subscriptor
```

File: examples/step-88/step-88.cc [hunk 21]
```diff
@@ -0,0 +1,898 @@
...
+    using FECellIntegrator = FEEvaluation<dim, -1, 0, 1, number>;
+
+    void reinit(const Mapping<dim>              &mapping,
+                const DoFHandler<dim>           &dof_handler,
+                const AffineConstraints<number> &constraints,
+                const Quadrature<dim>           &quad);
+
+    types::global_dof_index m() const;
+
+    number el(unsigned int, unsigned int) const;
```
Comment (bangerth) [inline examples/step-88/step-88.cc:201]: Do you actually need this function? If so, please make the arguments `const`.

File: examples/step-88/step-88.cc [hunk 22]
```diff
@@ -0,0 +1,898 @@
...
+
+    void initialize_dof_vector(VectorType &vec) const;
+
+    void vmult(VectorType &dst, const VectorType &src) const;
+
+    void Tvmult(VectorType &dst, const VectorType &src) const;
+
+    void compute_inverse_diagonal(VectorType &diagonal) const;
+
+    const TrilinosWrappers::SparseMatrix &get_system_matrix() const;
```
Comment (bangerth) [inline examples/step-88/step-88.cc:211]: Perhaps I was too quick in the comment about the class above in saying that it is a matrix-free implementation. What do you need the matrix for?

File: examples/step-88/step-88.cc [hunk 23]
```diff
@@ -0,0 +1,898 @@
...
+  {
+    return matrix_free.get_dof_handler().n_dofs();
+  }
+
+
+
+  template <int dim, typename number>
+  number LaplaceOperator<dim, number>::el(unsigned int, unsigned int) const
+  {
+    Assert(false, ExcNotImplemented());
```
Comment (bangerth) [inline examples/step-88/step-88.cc:262]:
```suggestion
    DEAL_II_NOT_IMPLEMENTED();
```

File: examples/step-88/step-88.cc [hunk 24]
```diff
@@ -0,0 +1,898 @@
...
+    return 0;
+  }
```
Comment (bangerth) [inline examples/step-88/step-88.cc:264]: Where is this function actually called? In other words, which place needs this function to exist, even though it isn't implemented?

File: examples/step-88/step-88.cc [hunk 25]
```diff
@@ -0,0 +1,898 @@
...
+    VectorType &diagonal) const
+  {
+    this->matrix_free.initialize_dof_vector(diagonal);
+    MatrixFreeTools::compute_diagonal(matrix_free,
+                                      diagonal,
+                                      &LaplaceOperator::do_cell_integral_local,
+                                      this);
+
+    for (auto &i : diagonal)
+      i = (std::abs(i) > 1.0e-10) ? (1.0 / i) : 1.0;
```
Comment (bangerth) [inline examples/step-88/step-88.cc:307]: Is this what is done in other programs as well? `1e-10` is an absolute number. If you're solving a temperature equation, then the units of matrix entries are typically of the kind "conductivity times meters^dimension / meters^2" and so will scale with physical parameters and the domain size. So if you say `1e-10` then I would ask "`1e-10` what?" Implementations that are robust with regard to physical parameters and domain sizes would make these sorts of criteria *relative* to something.

File: examples/step-88/step-88.cc [hunk 26]
```diff
@@ -0,0 +1,898 @@
...
+  }
+
+
+
+  template <int dim, typename number>
+  const TrilinosWrappers::SparseMatrix &
+  LaplaceOperator<dim, number>::get_system_matrix() const
+  {
+    if (system_matrix.m() == 0 && system_matrix.n() == 0)
```
Comment (bangerth) [inline examples/step-88/step-88.cc:316]: Did you mean `||` instead of `&&`?

File: examples/step-88/doc/intro.dox (hunk 3 above)
Comment (bangerth) [reply examples/step-88/doc/intro.dox:101]: Also add "with unit conductivity" to the description of the equation.

File: examples/step-88/step-88.cc [hunk 27]
```diff
@@ -0,0 +1,898 @@
...
+        integrator.reinit(cell);
+
+        do_cell_integral_global(integrator, dst, src);
+      }
+  }
+
+
+
+  template <int dim, typename number>
+  void LaplaceOperator<dim, number>::do_cell_integral_global(
```
Comment (bangerth) [inline examples/step-88/step-88.cc:387]: You also have the same function name with `_local` at the end. Are these standard names? One of these functions computes a matrix, the other computes a matrix-vector product, but that is not obvious from the function names.

File: examples/step-88/step-88.cc [hunk 28]
```diff
@@ -0,0 +1,898 @@
...
+    integrator.integrate_scatter(EvaluationFlags::gradients, dst);
+  }
+
+
+
+  // @sect3{The <code>Step88</code> class template}
+
+  // The main class has the same structure as in all other tutorial
+  // programs, with the obvious exception that no matrix needs to be assembled
+  // in this case. Notice that the create_grids() function return a Boolean that
```
Comment (bangerth) [inline examples/step-88/step-88.cc:406]: But the class above has the ability to assemble a matrix. Is that actually needed then?

Comment (bangerth) [inline examples/step-88/step-88.cc:406]:
```suggestion
  // programs, with the obvious exception that no matrix needs to be assembled
  // in this case. Notice that the `create_grids()` function return a Boolean that
```

File: examples/step-88/step-88.cc [hunk 29]
```diff
@@ -0,0 +1,898 @@
...
+    // The following objects are used to store, on every level of the hierarchy,
+    // its associated DoFHandler, AffineConstraints and the underlying
+    // differential operator.
+    MGLevelObject<DoFHandler<dim>>              dof_handlers;
+    MGLevelObject<AffineConstraints<Number>>    constraints;
+    MGLevelObject<LaplaceOperator<dim, Number>> operators;
+
+    // The next vectors store right-hand-side and solution.
+    VectorType rhs;
+    VectorType solution;
```
Comment (bangerth) [inline examples/step-88/step-88.cc:455]: This is much easier to read: comment blocks comment on multiple lines of code at once.

File: examples/step-88/step-88.cc [hunk 30]
```diff
@@ -0,0 +1,898 @@
...
+    const MPI_Comm     comm;
+    const Parameters   params;
+    const unsigned int min_level;
+    const unsigned int max_level;
+    ConditionalOStream pcout;
+
+    // Each level is described by an independent Triangulation object that will
+    // have to be accessed during the setup phase of the multigrid method.
+    // Hence, we store them in a standard vector of shared pointers.
+    std::vector<std::shared_ptr<Triangulation<dim>>> triangulations;
```
Comment (bangerth) [inline examples/step-88/step-88.cc:440]: Could these be `std::unique_ptr`s?

File: examples/step-88/step-88.cc [hunk 31]
```diff
@@ -0,0 +1,898 @@
...
+            auto triangulation =
+              std::make_shared<parallel::shared::Triangulation<dim>>(comm);
+            GridGenerator::convert_hypercube_to_simplex_mesh(dummy,
+                                                             *triangulation);
+            triangulation->refine_global(l);
+            triangulations.push_back(triangulation);
+          }
+
+        return true;
+      }
```
Comment (bangerth) [inline examples/step-88/step-88.cc:519]: If you already have this code in the program, it would be nice to also show results for this case (and mention it in the introduction).

File: examples/step-88/step-88.cc [hunk 32]
```diff
@@ -0,0 +1,898 @@
...
+    else if (params.mesh_type == "hyper_cube_with_simplices")
+      {
+        pcout << " - hyper_cube_with_simplices" << std::endl;
+        Triangulation<dim> dummy;
+        GridGenerator::hyper_cube(dummy);
+
+        for (unsigned int l = min_level; l <= max_level; ++l)
+          {
+            auto triangulation =
+              std::make_shared<parallel::shared::Triangulation<dim>>(comm);
```
Comment (bangerth) [inline examples/step-88/step-88.cc:511]: I think the function's overall documentation could also use a sentence that explains why you use 3 different kinds of triangulations :-)

File: examples/step-88/step-88.cc [hunk 33]
```diff
@@ -0,0 +1,898 @@
...
+            const TriangulationDescription::Description<dim> description =
+              TriangulationDescription::Utilities::
+                create_description_from_triangulation_in_groups<dim, dim>(
+                  [mesh_file_name](Triangulation<dim> &tria_base) {
+                    GridIn<dim> grid_in;
+                    grid_in.attach_triangulation(tria_base);
+                    grid_in.read(mesh_file_name, GridIn<dim>::abaqus);
+                  },
+                  [](Triangulation<dim> &tria_base,
+                     const auto          comm,
```
Comment (bangerth) [inline examples/step-88/step-88.cc:545]: Let's give this variable a type:

```suggestion
                     const MPI_Comm         mpi_communicator,
```

File: examples/step-88/step-88.cc [hunk 34]
```diff
@@ -0,0 +1,898 @@
...
+                     const auto) {
```
Comment (bangerth) [inline examples/step-88/step-88.cc:546]:
```suggestion
                     const unsigned int group_size) {
           AssertDimension(group_size == Utilities::MPI::n_mpi_processes(comm));
```

File: examples/step-88/step-88.cc [hunk 35]
```diff
@@ -0,0 +1,898 @@
...
+                    GridTools::partition_triangulation(
+                      Utilities::MPI::n_mpi_processes(comm), tria_base);
```
Comment (bangerth) [inline examples/step-88/step-88.cc:548]:
```suggestion
                      group_size, tria_base);
```

File: examples/step-88/step-88.cc [hunk 36]
```diff
@@ -0,0 +1,898 @@
...
+                  Utilities::MPI::n_mpi_processes(comm));
+
+            auto triangulation =
+              std::make_shared<parallel::fullydistributed::Triangulation<dim>>(
+                comm);
+            triangulation->create_triangulation(description);
+            triangulations.push_back(triangulation);
+          }
+
+        for (unsigned int l = max_level; l != 0; --l)
```
Comment (bangerth) [inline examples/step-88/step-88.cc:560]:
```suggestion
        for (unsigned int l = max_level; l>0; --l)
```

File: examples/step-88/step-88.cc [hunk 37]
```diff
@@ -0,0 +1,898 @@
...
+          {
+            const RepartitioningPolicyTools::ImmersedMeshPolicy<dim>
+              repartitioning_policy(*triangulations[l]);
+
+            const auto partition_vector =
+              repartitioning_policy.partition(*triangulations[l - 1]);
```
Comment (bangerth) [inline examples/step-88/step-88.cc:566]: This is actually pretty cool -- I didn't realize we had this facility. Would you mind adding in the description above that you're re-partitioning each triangulation in such a way that it matches the *next finer* triangulation? I assume you do this because it minimizes communication, right? Perhaps that reason is worth mentioning as well.

File: examples/step-88/step-88.cc [hunk 38]
```diff
@@ -0,0 +1,898 @@
...
+
+            const auto description = TriangulationDescription::Utilities::
+              create_description_from_triangulation(*triangulations[l - 1],
+                                                    partition_vector);
```
Comment (bangerth) [inline examples/step-88/step-88.cc:570]: Does that mean that you just overwrite the partition information you created for levels 0...L-1 above? Would a better scheme just not partition these lower-level grids at all if you don't need this information? Perhaps one could just read meshes fine-to-coarse, partition the finest one individually, and the other ones like you do here.

File: examples/step-88/step-88.cc [hunk 39]
```diff
@@ -0,0 +1,898 @@
...
+
+        pcout << std::endl;
+
+        return false; // non-nested mesh
+      }
+    else if (params.mesh_type == "gmesh_journal")
+      {
+        AssertThrow(false, ExcNotImplemented()); // TODO: ?
+        return false;
+      }
```
Comment (bangerth) [inline examples/step-88/step-88.cc:587]: If you don't need the code, then remove it (and also remove the "gmesh_journal" option from the input file parameter description).

File: examples/step-88/step-88.cc [hunk 40]
```diff
@@ -0,0 +1,898 @@
...
+
+    AssertThrow(false, ExcNotImplemented());
```
Comment (bangerth) [inline examples/step-88/step-88.cc:589]:
```suggestion
else
{
    DEAL_II_NOT_IMPLEMENTED
```

File: examples/step-88/step-88.cc [hunk 41]
```diff
@@ -0,0 +1,898 @@
...
+    // Next, we create finite element, mappings, and quadrature rules depending
+    // on the mesh type. In the case of hyper-cube meshes, FE_Q, MappingQ, and
+    // QGauss are used; in the case of simplex meshes, FE_SimplexP, MappingFE,
+    // and QGaussSimplex are used.
+    if (triangulations.back()->all_reference_cells_are_hyper_cube())
+      fe = std::make_unique<FE_Q<dim>>(params.fe_degree);
+    else if (triangulations.back()->all_reference_cells_are_simplex())
+      fe = std::make_unique<FE_SimplexP<dim>>(params.fe_degree);
+    else
+      AssertThrow(false, ExcNotImplemented());
```
Comment (bangerth) [inline examples/step-88/step-88.cc:610]: Perhaps combine the first in-function comment with the overall function comment into a single block (with an empty line/paragraph break in between) to keep some of the code flow better visible?

Also, starting a paragraph with "Next" means that the previous paragraph was a "first" or "second" :-)

Comment (bangerth) [reply examples/step-88/step-88.cc:610]:
```suggestion
  // After generating all levels, we can define on each one of them a suitable
  // FiniteElement space and distribute the DoFs. In particular, in this
  // tutorial we will consider classical Lagrangian finite element spaces.
  template <int dim>
  void LaplaceProblem<dim>::setup_system()
  {
    // Next, we create finite element, mappings, and quadrature rules depending
    // on the mesh type. In the case of hyper-cube meshes, FE_Q, MappingQ, and
    // QGauss are used; in the case of simplex meshes, FE_SimplexP, MappingFE,
    // and QGaussSimplex are used.
    if (triangulations.back()->all_reference_cells_are_hyper_cube())
      fe = std::make_unique<FE_Q<dim>>(params.fe_degree);
    else if (triangulations.back()->all_reference_cells_are_simplex())
      fe = std::make_unique<FE_SimplexP<dim>>(params.fe_degree);
    else
      DEAL_II_NOT_IMPLEMENTED();
```

File: examples/step-88/step-88.cc [hunk 42]
```diff
@@ -0,0 +1,898 @@
...
+
+    const auto reference_cell = triangulations.back()->get_reference_cells()[0];
```
Comment (bangerth) [inline examples/step-88/step-88.cc:612]:
```suggestion
    const ReferenceCell<dim> reference_cell = triangulations.back()->get_reference_cells()[0];
```

File: examples/step-88/step-88.cc [hunk 43]
```diff
@@ -0,0 +1,898 @@
...
+
+    // Initialize system of levels. This includes the setup
+    // of the corresponding DoFHandler, AffineConstraints, and
+    // LaplaceOperator objects.
+    dof_handlers.resize(min_level, max_level);
+    constraints.resize(min_level, max_level);
+    operators.resize(min_level, max_level);
+
+    // Then, we loop over all levels of the hierarchy and define on each one of
+    // them the associated operator.
```
Comment (bangerth) [inline examples/step-88/step-88.cc:626]:
```suggestion
    // Then, we loop over all levels of the hierarchy and define on each level
    // the associated operator.
```

File: examples/step-88/step-88.cc [hunk 44]
```diff
@@ -0,0 +1,898 @@
...
+                                        constraints.back());
+
+    pcout << std::endl;
+  }
+
+
+
+  // @sect4{Step88::solve}
+  // We are now ready to initialize the intergrid transfer operators,
+  // smoothers and preconditioner needed by the Multigrid infrastructure.
```
Comment (bangerth) [inline examples/step-88/step-88.cc:672]:
```suggestion
  // smoothers, and preconditioners needed by the Multigrid infrastructure.
```

File: examples/step-88/step-88.cc [hunk 45]
```diff
@@ -0,0 +1,898 @@
...
+    using SmootherType               = PreconditionChebyshev<LevelMatrixType,
+                                               VectorType,
+                                               SmootherPreconditionerType>;
+    using MGTransferType             = MGTransferMF<dim, Number>;
+    using PreconditionerType = PreconditionMG<dim, VectorType, MGTransferType>;
+
+    // Initialize multigrid transfer operator. For this purpose, we initialize
+    // two-level transfer operators between each neighboring level. In
+    // the non-nested case, we use MGTwoLevelTransferNonNested and, in the
+    // case of the nested case, we use MGTwoLevelTransfer. Both classes
```
Comment (bangerth) [inline examples/step-88/step-88.cc:688]:
```suggestion
    // case of nested meshes, we use MGTwoLevelTransfer. Both classes
```

File: examples/step-88/step-88.cc [hunk 46]
```diff
@@ -0,0 +1,898 @@
...
+        smoother_data[l].smoothing_range     = 20;
+        smoother_data[l].degree              = 5;
+        smoother_data[l].eig_cg_n_iterations = 20;
+      }
+
+    MGSmootherPrecondition<LevelMatrixType, SmootherType, VectorType>
+      mg_smoother;
+    mg_smoother.initialize(operators, smoother_data);
+
+    // As coarse-grid solver, we use a single v-cycle of AMG.
```
Comment (bangerth) [inline examples/step-88/step-88.cc:761]:
```suggestion
    // As coarse-grid solver, we use a single V-cycle of AMG.
```

File: examples/step-88/step-88.cc [hunk 47]
```diff
@@ -0,0 +1,898 @@
...
+    MGCoarseGridApplyOperator<VectorType, TrilinosWrappers::PreconditionAMG>
+      mg_coarse(precondition_amg);
+
+    // Finally, we can initialize the Multigrid object and use it to
+    // precondition our conjugate-gradient solver. Parameters related to the
+    // CG algorithm are read from the parameter files.
+    Multigrid<VectorType> mg(
+      mg_matrix, mg_coarse, mg_transfer, mg_smoother, mg_smoother);
+
+    PreconditionerType preconditioner(dof_handlers.back(), mg, mg_transfer);
```
Comment (bangerth) [inline examples/step-88/step-88.cc:774]: It's been nearly 100 lines since `PreconditionerType` was declared. I had to scroll way up to get there to look up what that was. For types that you use only once, perhaps it's useful to not introduce these abbreviations but just insert the type right here.

File: examples/step-88/step-88.cc [hunk 48]
```diff
@@ -0,0 +1,898 @@
...
+
+        Vector<float> ranks(triangulations[l]->n_active_cells());
+        ranks = Utilities::MPI::this_mpi_process(comm);
+        data_out.add_data_vector(ranks, "ranks");
+
+        data_out.build_patches();
+        data_out.write_vtu_in_parallel("grid_" + std::to_string(l) + ".vtu",
+                                       comm);
+      }
+  }
```
Comment (bangerth) [inline examples/step-88/step-88.cc:825]: Would it be useful to visualize the partitioning of these meshes in the results section?

File: examples/step-88/step-88.cc [hunk 49]
```diff
@@ -0,0 +1,898 @@
...
+
+
+
+  // The run function simply calls all other methods in the
```
Comment (bangerth) [inline examples/step-88/step-88.cc:829]: Add a `@sect` tag for this function so it shows up in the table of contents.

File: examples/step-88/step-88.cc [hunk 50]
```diff
@@ -0,0 +1,898 @@
...
+    setup_system();
+    solve();
+    output_results();
+  }
+
+} // namespace Step88
+
+
+
+int main(int argc, char *argv[])
```
Comment (bangerth) [inline examples/step-88/step-88.cc:847]: Same here, plus a one-line comment.

**Review by bangerth — COMMENTED** [review]
Here's the rest. Really nice program! <3

**Comment by bangerth** [comment]
@fdrmrc Ping?

**Comment by bangerth** [comment]
I'm afraid we've run out of time on this one. Let's postpone till 9.9 unless we get the comments resolved in the next couple of days.
